drivers/src/pmu.rs

```diff
@@ -50,19 +53,62 @@ impl PmuInfo {
         Ok(ControlFlow::Continue(()))
     }
 
+    fn init_delegated_counters(&mut self) -> SbiResult<()> {
+        use riscv_regs::CSR;
+        // CYCLE and INSRET must be delegated
+        const MANDATORY_DELEGATED_COUNTERS_MASK: u64 = 0x5;
+        let delegated_counters = CSR.smcounteren.get();
+        if (delegated_counters & MANDATORY_DELEGATED_COUNTERS_MASK) == 0 {
+            return Err(SbiError::NotSupported);
+        }
+
+        self.valid_hw_counters_mask = delegated_counters;
+        for i in 0..(MAX_HARDWARE_COUNTERS as u64) {
+            if (delegated_counters & (1 << i)) != 0 {
+                let base_csr = riscv_regs::CSR_CYCLE as u64;
+                let info = PmuCounterInfo::from_csr(base_csr + i);
+                self.hw_counters_info.push(Some(info));
+            } else {
+                // Hardware counters can be sparse, and may not be implemented.
+                self.hw_counters_info.push(None);
+            }
+        }
+
+        Ok(())
+    }
+
     /// Initializes the global PmuInfo structure with the information obtained from the platform SBI.
     pub fn init() -> SbiResult<()> {
         let mut pmu_info = PmuInfo::new();
-        let num_counters = pmu::get_num_counters()?;
-        for i in 0..num_counters.min(MAX_HARDWARE_COUNTERS as u64) {
-            if matches!(pmu_info.init_counter(i)?, ControlFlow::Break(())) {
-                break;
+        let cpu_id = CpuInfo::get();
+        if !cpu_id.has_sscdeleg() {
+            let num_counters = pmu::get_num_counters()?;
```

Comment:
IMO, it would be nice to split into different functions fro code quality perspective. It makes it easier to read through.
Something like this:

get_num_counters_sbi() {
}

get_delegated_counters() {

}

e.g.
pmu::get_num_counters() {
     if (cpu_id.has_sscdeleg())
           get_delegated_counters()
      else
           get_num_counters_sbi()
}

Same goes for stopping counters below and all other counter related ops.